File: IsRecursivo.java
/*
- Refaça a questão Is de forma recursiva.
- 4 métodos: verifica se a string é só vogais, só consoantes, um número
  inteiro, ou um número real. Saída: X1 X2 X3 X4 (SIM/NAO).
*/
import java.util.Scanner;

public class IsRecursivo {

    /**
     * Método público que inicia a recursão para verificar se a string
     * é composta somente por vogais.
     * @param texto a string a verificar
     * @return true se todos os caracteres forem vogais
    */

    public static boolean somenteVogais(String texto) {
        return somenteVogaisRec(texto, 0);
    }

    public static boolean somenteVogaisRec(String texto, int i) {
        if (i == texto.length()) { // caso base: fim da string
            return true;
        }
        char c = texto.charAt(i);
        boolean ehVogal = (c=='a'||c=='e'||c=='i'||c=='o'||c=='u'||
                           c=='A'||c=='E'||c=='I'||c=='O'||c=='U');
        if (!ehVogal) {
            return false;
        }
        return somenteVogaisRec(texto, i + 1);
    }

    /**
     * Método público que inicia a recursão para verificar se a string
     * é composta somente por consoantes.
     * @param texto a string a verificar
     * @return true se todos os caracteres forem consoantes
    */

    public static boolean somenteConsoantes(String texto) {
        return somenteConsoantesRec(texto, 0);
    }

    public static boolean somenteConsoantesRec(String texto, int i) {
        if (i == texto.length()) {
            return true;
        }
        char c = texto.charAt(i);
        boolean ehLetra = (c>='a' && c<='z') || (c>='A' && c<='Z');
        boolean ehVogal = (c=='a'||c=='e'||c=='i'||c=='o'||c=='u'||
                           c=='A'||c=='E'||c=='I'||c=='O'||c=='U');
        if (!ehLetra || ehVogal) {
            return false;
        }
        return somenteConsoantesRec(texto, i + 1);
    }

    /**
     * Método público que inicia a recursão para verificar se a string
     * representa um número inteiro válido.
     * @param texto a string a verificar
     * @return true se for um número inteiro válido
    */

    public static boolean ehInteiro(String texto) {
        if (texto.length() == 0) {
            return false;
        }
        return ehInteiroRec(texto, 0);
    }

    public static boolean ehInteiroRec(String texto, int i) {
        if (i == texto.length()) {
            return true;
        }
        char c = texto.charAt(i);
        if (!(c >= '0' && c <= '9')) {
            return false;
        }
        return ehInteiroRec(texto, i + 1);
    }

    /**
     * Método público que verifica se a string representa um número real
     * válido: precisa ter exatamente 1 ponto e o restante só dígitos.
     * @param texto a string a verificar
     * @return true se for um número real válido
    */
   
    public static boolean ehReal(String texto) {
        if (texto.length() == 0) {
            return false;
        }
        return contarPontos(texto, 0) == 1 && soDigitoOuPonto(texto, 0);
    }

    // conta os pontos somando 1 (se achar '.') ou 0 com o resultado do restante da string
    public static int contarPontos(String texto, int i) {
        if (i == texto.length()) {
            return 0;
        }
        int atual = (texto.charAt(i) == '.') ? 1 : 0;
        return atual + contarPontos(texto, i + 1);
    }

    public static boolean soDigitoOuPonto(String texto, int i) {
        if (i == texto.length()) {
            return true;
        }
        char c = texto.charAt(i);
        if (!((c >= '0' && c <= '9') || c == '.')) {
            return false;
        }
        return soDigitoOuPonto(texto, i + 1);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        while (scanner.hasNextLine()) {
            String linha = scanner.nextLine();
            String x1 = somenteVogais(linha) ? "SIM" : "NAO";
            String x2 = somenteConsoantes(linha) ? "SIM" : "NAO";
            String x3 = ehInteiro(linha) ? "SIM" : "NAO";
            String x4 = ehReal(linha) ? "SIM" : "NAO";
            System.out.println(x1 + " " + x2 + " " + x3 + " " + x4);
        }
    }
}

File: IsRecursivo.c
/*
- Refaça a questão Is de forma recursiva.
- 4 métodos: verifica se a string é só vogais, só consoantes, um número
  inteiro, ou um número real. Saída: X1 X2 X3 X4 (SIM/NAO).
*/

#include <stdio.h>
#include <stdlib.h>

int minhaStrlen(char* s){
    int contador = 0;
    while(s[contador] != '\0'){
        contador++;
    }
    return contador;
}

/**
 * Verifica, recursivamente, se todos os caracteres a partir da posição i
 * são vogais.
 * @param texto a string a verificar
 * @param i a posição atual sendo checada
 * @return 1 se todos forem vogais, 0 caso contrário
 */
int somenteVogaisRec(char* texto, int i){
    if (i == minhaStrlen(texto)) { // caso base: fim da string
        return 1;
    }
    char c = texto[i];
    int ehVogal = (c=='a'||c=='e'||c=='i'||c=='o'||c=='u'||
                   c=='A'||c=='E'||c=='I'||c=='O'||c=='U');
    if (!ehVogal) {
        return 0;
    }
    return somenteVogaisRec(texto, i + 1);
}

/**
 * Método público que inicia a recursão a partir da posição 0.
 * @param texto a string a verificar
 * @return 1 se todos os caracteres forem vogais
 */
int somenteVogais(char* texto){
    return somenteVogaisRec(texto, 0);
}

/**
 * Verifica, recursivamente, se todos os caracteres a partir da posição i
 * são consoantes.
 * @param texto a string a verificar
 * @param i a posição atual sendo checada
 * @return 1 se todos forem consoantes, 0 caso contrário
 */
int somenteConsoantesRec(char* texto, int i){
    if (i == minhaStrlen(texto)) {
        return 1;
    }
    char c = texto[i];
    int ehLetra = (c>='a' && c<='z') || (c>='A' && c<='Z');
    int ehVogal = (c=='a'||c=='e'||c=='i'||c=='o'||c=='u'||
                   c=='A'||c=='E'||c=='I'||c=='O'||c=='U');
    if (!ehLetra || ehVogal) {
        return 0;
    }
    return somenteConsoantesRec(texto, i + 1);
}

int somenteConsoantes(char* texto){
    return somenteConsoantesRec(texto, 0);
}

/**
 * Verifica, recursivamente, se todos os caracteres a partir da posição i
 * são dígitos.
 * @param texto a string a verificar
 * @param i a posição atual sendo checada
 * @return 1 se todos forem dígitos, 0 caso contrário
 */
int ehInteiroRec(char* texto, int i){
    if (i == minhaStrlen(texto)) {
        return 1;
    }
    char c = texto[i];
    if (!(c >= '0' && c <= '9')) {
        return 0;
    }
    return ehInteiroRec(texto, i + 1);
}

int ehInteiro(char* texto){
    if (minhaStrlen(texto) == 0) { // string vazia não é número válido
        return 0;
    }
    return ehInteiroRec(texto, 0);
}

/**
 * Conta, recursivamente, quantos caracteres '.' existem a partir de i.
 * Soma o valor da posição atual (0 ou 1) com o resultado da chamada
 * recursiva, sem usar parâmetro acumulador.
 * @param texto a string a verificar
 * @param i a posição atual sendo checada
 * @return quantidade de pontos encontrados a partir de i
 */
int contarPontos(char* texto, int i){
    if (i == minhaStrlen(texto)) {
        return 0;
    }
    int atual = (texto[i] == '.') ? 1 : 0;
    return atual + contarPontos(texto, i + 1);
}

/**
 * Verifica, recursivamente, se todos os caracteres a partir de i são
 * dígitos ou o ponto decimal.
 * @param texto a string a verificar
 * @param i a posição atual sendo checada
 * @return 1 se todos forem dígito ou ponto, 0 caso contrário
 */
int soDigitoOuPontoRec(char* texto, int i){
    if (i == minhaStrlen(texto)) {
        return 1;
    }
    char c = texto[i];
    if (!((c >= '0' && c <= '9') || c == '.')) {
        return 0;
    }
    return soDigitoOuPontoRec(texto, i + 1);
}

int ehReal(char* texto){
    if (minhaStrlen(texto) == 0) {
        return 0;
    }
    return contarPontos(texto, 0) == 1 && soDigitoOuPontoRec(texto, 0);
}

int main(){
    char linha[1000];
    while(fgets(linha, 1000, stdin) != NULL){
        int tamanho = minhaStrlen(linha);
        if(linha[tamanho - 1] == '\n'){
            linha[tamanho - 1] = '\0';
        }

        printf("%s ", somenteVogais(linha) ? "SIM" : "NAO");
        printf("%s ", somenteConsoantes(linha) ? "SIM" : "NAO");
        printf("%s ", ehInteiro(linha) ? "SIM" : "NAO");
        printf("%s\n", ehReal(linha) ? "SIM" : "NAO");
    }
    return 0;
}